### ISSUE TITLE
[TaskManager][DependencyManager] Spawn dependencies outside of the task manager

### ISSUE THREAD

**Issue by kdelee** [ansible/awx#12202]
##### ISSUE TYPE
 - Feature Idea

##### SUMMARY

Re: incremental task manager improvments

From @AlanCoding :
Spawn “hard” dependencies outside of the task manager, details:
On creation, job triggers a different task
That does miscellaneous work, including dependency spawning
That triggers the task manager
Task manager needs to know to not process these just-created jobs

How:
   We could continue using `dependencies_processed` field XOR Could leave jobs in “new” status until task is finished… already an option!

This would probably touch the queries in https://github.com/ansible/awx/blob/a86740c3c9eaf9a551e850341d8adec5a3962dd5/awx/main/scheduler/task_manager.py#L89-L101 

NOTE: we are also talking about modifying this area of the code in https://github.com/ansible/awx/issues/12199

Avoids spawning dependencies while holding global lock, can adopt other logic

**Comment by fosterseth** [comment]
Also potentially spawn workflow jobs from a workflow JT in the same or similar preparation task

**Comment by kdelee** [comment]
https://github.com/ansible/awx/blob/923cc671db1e8a76325afc6ed14798bcbc66a64d/awx/main/scheduler/task_manager.py#L151-L212

**Comment by AlanCoding** [comment]
Oof, not only does `bfs_nodes_to_run` make references to `p.job`, where `p` is a parent worthy of considering for whether next node should run, it also references `p.unified_job_template`. This has potential for virtually unlimited performance impact and it applies for every run of the task manager when a workflow is half-finished.

**Comment by kdelee** [comment]
@fosterseth and @TheRealHaoLiu are thinking it would be ideal of the DependencyManager doesn't "release" jobs to the TaskManager until hard and soft dependencies are resolved. This way the TaskManager could consider that all jobs its acting on are ready to go

"Soft Blocks"
- allow_simultaneous is false, we should probably start in "blocked" and only release if we are unblocked

@AlanCoding what are other "soft blocks" we need to consider?

**Comment by kdelee** [comment]
![Screenshot from 2022-07-01 16-14-32](https://user-images.githubusercontent.com/29239279/176962859-b4fb4453-76e5-4822-b0d4-b1f66930dd5a.png)

**Comment by AlanCoding** [comment]
That would mean that `DependencyManager` will process the same jobs multiple times. This is the fundamental nature of "soft" dependencies. Other forms of soft dependencies are preventing multiple project updates from running at the same time on the same project, or multiple inventory updates running on the same time on the same inventory. The core logic is mainly hosted here:

https://github.com/ansible/awx/blob/devel/awx/main/scheduler/dependency_graph.py

> ideal of the DependencyManager doesn't "release" jobs to the TaskManager until hard and soft dependencies are resolved

The pained part of this is that "soft" dependencies are never truly resolved. If a job gets unblocked due to its job template, it may then _get blocked_ by its project, or something along those lines. In other words, if you allow that the DependencyManager may re-process jobs according to their soft dependencies, you do not assure that the TaskManager will not _also_ re-process those jobs because of the same blocking reasons.

**Comment by kdelee** [comment]
@AlanCoding I think I may be ok with that. This would move the existing TaskManager closer to being the "CapacityManager" e.g. only concerned with "what has capacity to run this job" and less with all these blocking conditions. But I get that it just moves a problem around

**Comment by AlanCoding** [comment]
If you want separate classes for all of the conditions, a class for hard dependencies, a class for soft dependencies, and a class for capacity management, then that's fine for concept separation. But the current branch has each class run under a global lock specific to itself, and that's particularly unnecessary for hard dependencies.

The general problem of re-processing is solvable, but may not be desirable. Similar to what is proposed in https://github.com/ansible/awx/issues/12440, you can avoid repeated unnecessary inspection of the pending jobs list by having a code hook _triggered by an unblocking event_. If a JT with allow_simultaneous=False frees up (a job finishes running), then loop over the pending jobs waiting on that resource and advance any that don't have other active blockers. You sacrifice idempotence in doing this, and things could get lost, requiring cleanup tasks. The bigger problem that the "no other blockers" criteria could balloon to unbounded complexity, putting you right back to building the full graph every time.

Because of that, I think it still makes sense to have a specialist class to manage _all forms_ of soft dependencies, which includes the soft resource blocks, as well as capacity. Both are forms of ephemeral blocking, and TaskManger is just the name that we're left with for this function. If something gets unblocked by a JT freeing up, then you still need to run through the gambit of other conditional blockers... and capacity in particular. It's kind of interesting to consider separating these, but it's likely to stir up more dragons, particularly if there's not great clarity of thought going into it.

**Comment by fosterseth** [comment]
In order to cut out looping over pending jobs to see if their hard dependencies are done in TaskManager, we can delay updating dependencies_processed to True until the hard dependencies are finished running.

the dependencies have the opportunity to alter the job that spawn that dependency in their final_run_hook. The issue is a given job can have more than 1 dependency (project update, multiple inventory source updates), and so how do we aggregate that information to determine a final dependencies_processed = True state.

In the final_run_hook of a spawned dependency, we can check on the other dependencies for that the job, and if all are finished, we then set dependencies_processed = True. The task manager does this exact logic already, we would just move this to the final_run_hook of the UnifiedJob.

So a project update can check the p.unifiedjob_blocked_jobs.all() to get a list of jobs that are waiting on it, and check the dependent_jobs.all() for each of _those_ jobs, and if all are status in finished state, mark that job with `dependencies_processed = True`.

**Comment by AlanCoding** [comment]
I think those are all very good ideas @fosterseth. My only concern is that, if you change the meaning of `dependencies_processed` like that, then you lose your means for distinguishing jobs which have _spawned_ dependencies from those that have not. Since spawning is something which only happens once for each job, this is something we really want to keep separate from the other less-deterministic logic.

There are several ways you could move in the direction you mention without making sacrifices. I will offer one suggestion here, because I think it can help lower task manager tech debt.

The spawning methods in the task manager are like:
 - get_latest_inventory_update
 - get_latest_project_update
 - should_update_inventory_source
 - should_update_related_project
 - create_project_update
 - create_inventory_update
 - add_dependencies

These method names tell the story of the psuedo-code easily. The parent method is `gen_dep_for_job`, which obviously checks for running inventory/project updates. Depending on what it finds, it creates inventory/project updates(s) and adds them to the job's dependency list.

At most, this could create (1 project update) + (N inventory updates) where N is the number of inventory sources for the related inventory. This is all very simple, but our code also does a bunch of questionable stuff that we should axe, for example:

> Inventory created 2 seconds behind job

Why? I suspect we are still carrying hacks related to the re-processing of _unspawned_ dependencies, which predates the `dependent_jobs` relationship _and_ the `dependencies_processed` field, and I believe we don't need to concern ourselves with preserving these hacks.

To my ultimate point - why not spawn dependencies when you create the job, at POST to `/api/v2/job_templates/N/launch/` for example? You've already started loading more logic into that. If it's not performant, then a request is easier to optimize than a bunch of monolithic task manager task code. Creating, maybe, 5 updates as a side effect doesn't seem intractable to me.

**Comment by kdelee** [comment]
the core work of creating the DependencyManager has already been merged into the task_manager_refactor branch -- going to close this and we can open up follow on issues
